Overhaul lesson + test suite

gj opened the pull request:
Fixes #33 
Fixes #32 
Fixes #30 
Fixes #29 
Fixes #25 
Fixes #24 
Fixes #23 

Closes #34 
Closes #28 
Closes #27

gj committed:
remove trailing spaces

gj committed:
Fix minor typos and grammar issues

gj committed:
Add instructions for why students need to create Concerns::Findable first (#24)

gj committed:
Add secondary condition to test sweeper to ensure that classes for which .destroy_all has not yet been defined also get purged (#33)

gj committed:
Improve test coverage and organization for the basic functionality of the Song, Artist, and Genre classes

gj committed:
Reorganize and rewrite the tests for the Song <-> Artist association to better cohere them with the lab instructions

gj committed:
Reorganize and rewrite the tests for the Song <-> Genre association to better cohere them with the lab instructions

gj committed:
Increase test accuracy by using the 'to be' matcher instead of 'to eq'

gj committed:
Rework Artist#add_song tests to make them less convoluted

gj committed:
Use '.to include' matcher instead of '.to be' for a more intuitive failure message

gj committed:
Fix minor issues in the test suite for the Song <-> Genre association

gj committed:
Change minor details in the descriptions and code of the Artist#add_song tests to further simplify them

gj committed:
Simplify and standardize tests for the Artist <-> Genre association

gj committed:
add additional test coverage to the Findable class methods on Song

gj committed:
Expand and improve test coverage for the Concerns::Findable module (#24 #32)

gj committed:
Expand and reorganize the Concerns::Findable test suite to reduce redundancy and improve coverage (#24 #32)

gj committed:
Narrow focus of MusicImporter test suite and add test to ensure filenames are sorted alphabetically in MusicImporter#import (#23 #25 #30)

gj committed:
Expand and enhance isolation and efficiency of MusicLibraryController test suite; fix rspec --fail-fast error

gj committed:
Reorganize the MusicLibraryController CLI tests and add a test for #play_song to ensure that entering a 0 or negative number does not play a song

gj committed:
Update Gemfile

gj committed:
Add explanatory note and tests to ensure students stop using @artist + @genre instance variables in favor of calling Song#artist= and Song#genre= in Song#initialize

gj committed:
Test Concerns::Findable methods against real classes instead of a TestClass
to not confuse students with too much isolated unit testing at this early point in the curriculum
and removes now-empty 009_findable_artists_and_genres_spec.rb, shifting every other test filename down by one

gj committed:
Shift hard-coding test coverage away from using File.read and instead rely on a second test library of 'songs'
to make test suite less brittle

gj committed:
Improve organization of test suite by refactoring into lets and spacing everything out

aviflombaum commented on README.md:
```diff
@@ -1,34 +1,42 @@
-This is a complex lab with many parts, go slow, try to understand what you're trying to build holistically before starting. Read this entire README before jumping in. As you go from spec to spec, we recommend doing them in numbered order. 
+This is a complex lab with many parts, so go slowly. Take time to understand what you're building holistically before starting. Read this entire README before jumping in. As you go from spec to spec, we recommend doing them in numbered order.
 
 ## Concerns
```
I would just include this in the original lab - can only create confusion and it's a stub anyway, we can leave the note in, but make it easy for them.

aviflombaum commented on README.md:
````diff
@@ -1,34 +1,42 @@
 ```
 
-Additionally, each class should contain a class variable `@@all` that is set to an empty array and is prepared to store all saved instances. This class variable should be accessible via the class method `.all`.
+Additionally, *each* class should contain a class variable `@@all` that is set to an empty array and is prepared to store all saved instances of the class. This class variable should be accessible via the class method `.all`.
````
This feels like too much of a hint - we really want to say that the class should expose all instantiated instances via the class method all, no?

aviflombaum commented on README.md:
````diff
@@ -1,34 +1,42 @@
 ```
 
-Instances should respond to a `#save` method that adds the instance itself into the class variable `@@all`.
+Instances should respond to a `#save` method that adds the instance itself into the appropriate `@@all` class variable.
````
Ya, there are tests that describe this so maybe we need to sort of make these instructions correspond to the tests, almost as like hints or something.

aviflombaum commented on README.md:
```diff
@@ -38,13 +46,13 @@ Song.all #=> [#<Song: @name="Blank Space">]
-Song.all #=> [#<Song: @name="Blank Space">]
+Song.new("Kaohsiung Christmas").save
+Song.all #=> [#<Song: @name="Blank Space">, #<Song: @name="Kaohsiung Christmas">]
 Song.destroy_all
```
That comes out of nowhere - we should mention that a `destroy_all` method should forget or clear the collection of instances.

aviflombaum commented on README.md:
```diff
@@ -55,64 +63,58 @@ Song.all #=> [#<Song: @name="Blank Space">]
 
-### Songs and Artists
+### `Song`s and `Artist`s
+ * Songs belong to an artist and an artist has many songs. Adding a song to an artist is done by calling an `#add_song` method on an instance of the `Artist` class.
```
This language of belongs_to and has_many, has it been introduced? Otherwise might be a nice time to mention the convention and describe it a bit.

jmburges commented on README.md:
```diff
@@ -55,64 +63,58 @@ Song.all #=> [#<Song: @name="Blank Space">]
-  * Genres have many songs and are initialized with an empty list of songs
-  * Songs have one genre
-  * Songs can be initialized with an optional genre
+***Note:*** when you reach the tests concerned with migrating your `Song#initialize` method away from setting instance variables for `@artist` and `@genre` and toward using the custom setter methods that you will define (e.g., `Song#genre=`), ensure that those setter methods are only invoked _if_ `Song#initialize` is called with artist and/or genre arguments. If we call `Song.new("Song Title", ar ...
```
We gotta put in some "why" here.

jmburges commented on README.md:
```diff
@@ -55,64 +63,58 @@ Song.all #=> [#<Song: @name="Blank Space">]
+## It's CLI time!
+Congrats! You've done the heavy lifting. Now let's wrap it all up in a simple CLI so that users can actually interact with our code.
+  * Upon initialization, the CLI should accept an optional path to the library of MP3 files, defaulting to `./db/mp3s/`. It should then instantiate a `MusicImporter` object, which it will use to import songs from the specified library.
+  * Add a `#call` method that starts the CLI and prompts the user for input. Read the tests carefully for specifics.
```
I'm like 99% sure this is correct, but we have shown them the weirdo receives get and responds with stuff before right?

gj replied:
You mean the simulated interactions in the test suite?

gj replied:
`$stdout` covered here: https://github.com/learn-co-curriculum/ruby-gets-input

jmburges commented on spec/001_song_basics_spec.rb:
```diff
@@ -1,54 +1,68 @@
+
+  describe ".destroy_all" do
+    it "resets the @@all class variable to an empty array" do
       Song.destroy_all
```
Each describes should be separate. Set the `@@all` array to have something in it, then destroy, then check empty

aviflombaum commented on spec/001_song_basics_spec.rb:
```diff
@@ -1,54 +1,68 @@
-    it 'accepts a name for the song' do
-      song = Song.new("In the Aeroplane Over the Sea")
+describe "Song" do
+  let(:song) { Song.new("In the Aeroplane Over the Sea") }
```
Ya, so I intentionally was avoiding using `let` in these early labs...we can discuss. I want to keep the narrative within the `it` block

gj replied:
Ended up keeping the `let`s but overriding them for the first couple tests (until the student has a functioning `#initialize` method) so that the error messages never mention the `let`s.

aviflombaum commented on spec/001_song_basics_spec.rb:
```diff
@@ -1,54 +1,68 @@
+
+  describe "#initialize" do
+    it "accepts a name for the new song" do
+      expect(song.instance_variable_defined?(:@name)).to be(true)
```
Ya, I think this isn't actually testing what is required - initialize accepting the name should be via the reader...the order of the tests matter. Not to mention, from an error perspective - this is going to say "@name not defined" as opposed to `song doesn't respond to name` which I think is better...let's chat about the pedagogy here.

jmburges commented on spec/002_artist_basics_spec.rb:
```diff
@@ -1,54 +1,68 @@
+  describe ".destroy_all" do
+    it "resets the @@all class variable to an empty array" do
       Artist.destroy_all
+
```
Same as above.

jmburges commented on spec/003_genre_basics_spec.rb:
```diff
@@ -1,54 +1,68 @@
-    it 'returns the class variable @@all' do
-      Genre.class_variable_set(:@@all, [])
+  describe '@@all' do
+    it 'is set to an empty array' do
```
It's really that it is set to an empty array before any save command right?

gj replied:
`it "is initialized as an empty array" do`

?

aviflombaum commented:
I think a lot of this is good but I'm a bit concerned about some of the specificity and the order. Happy to review. You can choose to ignore or keep the changes.

jmburges commented on spec/004_songs_and_artists_spec.rb:
```diff
@@ -1,102 +1,101 @@
         song.artist = artist
 
-        expect(artist.songs).to include(song)
+        expect(song.artist).to be(artist)
```
I would check the instance variable here rather than re-calling `#artist`

jmburges commented on spec/007_findable_songs_spec.rb:
```diff
@@ -1,21 +1,35 @@
+    it "finds (does not recreate) an existing song with the provided name if one exists in @@all" do
+      same_song = Song.find_or_create_by_name("In the Aeroplane Over the Sea")
+
+      expect(same_song).to be(song)
```
So this is really just testing the return of the `find_or_create_by_name` function. Maybe include a double check the song.all method for only one thing.

gj replied:
Even though it's already checking object identity (`.to be`)?

jmburges commented on spec/007_findable_songs_spec.rb:
```diff
@@ -1,21 +1,35 @@
+    end
+
+    it "creates a song if an existing match is not found" do
+      song = Song.find_or_create_by_name("I'd Rather Go Blind")
```
Different variable name here than the `let`

jmburges commented on spec/009_music_importer_spec.rb:
```diff
@@ -0,0 +1,95 @@
+      music_importer.import
+    end
+
+    it "sorts the files alphabetically prior to creating new songs from the filenames" do
```
Why does it have to sort before creating? What do you actually care about here? Not that the sorting happens on input, just that it happens correctly on output. Test shouldn't care *when* the sort happens

gj replied:
You right you right. Fixed: cf83bf8

jmburges reviewed (approved):
Some small things, fix 'em and :shipit:

gj commented:
@aviflombaum specificity as in the tests don't leave enough wiggle room for students to solve the lab in unique ways? Would love to hear thoughts on how to better order everything.

aviflombaum commented:
Ya, that's what I mean by specificity, I also think about what the failing test error message will read and if it's providing as much information and clues to the student.

gj commented:
@aviflombaum anecdotally, I tried to use `.to eq` in place of `.to be` where object identity isn't super important b/c the error message generated by `.to eq` is much less daunting / verbose, but I'd love any tips on how to improve on that front

gj committed:
Make minor edits to some tests per @jmburges suggestions

gj committed:
Shift the burden of sorting alphabetically from MusicImporter#import to the MusicLibraryController display methods

gj committed:
Simplify first few tests for the Song, Artist, and Genre class basics
to maintain tests' focus on student comprehension (per @aviflombaum suggestion)

gj committed:
Remove obsolete section in README explaining hacky module definition steps to get test suite functioning

gj committed:
Check that Concerns::Findable is not a Class
because simply checking if it's a Module passes regardless of whether it's a Module or a Class

gj committed:
Make test failures a lot more readable when checking whether Concerns::Findable is properly extending Artist and Genre
